File: T2/optimize.c
#include "include/optimize.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int min_mem = 0;
int niveles[5] = {0};
int minimos[6] = {0};

int* optimize(int n, int bits, int mem_sum) {
    int x_mem = 0;
    int tot_mem = 0;

    if (n == 1) {
        if (bits == N_BITS) {
            tot_mem = mem_sum + (int)pow(2, bits) * 11;
        }
        else {
            tot_mem = mem_sum + (int)pow(2, bits) * niveles[1];
        }
    
        if (tot_mem < min_mem || min_mem == 0) {
            min_mem = tot_mem;
            niveles[n-1] = bits;
            for (int i=0; i < 5; i++) {
                minimos[i] = niveles[i];
            }
        }
        minimos[-1] = min_mem;
        return minimos;
    }

    for (int x=1; x < bits; x++) {
        niveles[n-1] = x;
        if (bits == N_BITS) {
            x_mem =  (int)pow(2, x) * 11;
        }
        else {
            x_mem =  (int)pow(2, x) * niveles[n];
        }
        optimize(n-1, bits-x, mem_sum + x_mem);
    }
    return minimos;
}

File: T2/optimize.py
N_BITS = 20
min_mem = 0
niveles = [0,0,0,0,0]
minimo = []

def optimize(n, bits, mem_sum=0):
    global min_mem
    global minimo
    if n == 1:
        if bits == N_BITS:
            tot_mem = mem_sum + ((2**bits) * 11)
        else:
            tot_mem = mem_sum + ((2**bits) * niveles[1])
        if tot_mem < min_mem or min_mem == 0:
            min_mem = tot_mem
            niveles[n-1] = bits
            minimo = [i for i in niveles]
        return
    for x in range(1, bits):
        niveles[n-1] = x
        if bits == N_BITS:
            x_mem = (2**x) * (11)
        else:
            x_mem = (2**x) * niveles[n]
        optimize(n-1, bits-x, mem_sum + x_mem)


for j in range(1,6):
    optimize(j, N_BITS)
    string = "Reparticion: "
    for i in minimo:
        string += "{}-".format(i)
    string += "bits\nMemoria usada {} bits".format(min_mem)
    print(string)
